Implement optional per-field encryption for wallet sqlite databases

philipr-za opened the pull request:
Apologies for the size of this PR, it was a difficult piece of work to split up into parts that were functional on their own.  A large portion of the PR is tests.

## Description
Currently sensitive data in the Wallet Sqlite databases are stored in the clear. On the mobile clients we rely on the security of the OS to restrict file access to these databases. 

This PR implements optional per-field encryption to allow for encrypting sensitive data before it is written to the sqlite db and decrypted when it is read from the sqlite db. AES256-GCM encryption is used for this process and the key is derived from a passphrase string that is hashed to a 256-bit key using Blake2. The nonce is stored with the cipher text in the database field.

An Encryptable trait is introduced which specifies the encrypt() and decrypt() functions that a struct should implement to encrypt and decrypt itself in-place with the provided cipher. All the Sql structs that have fields that should be encrypted/decrypted implement this trait and when database access is performed with a supplied Cipher these implementations are used.

In order to know if the database is currently encrypted and whether the provided Passphrase is correct a validation is performed using the Wallet Database. The wallet database persists the Comms Secret Key and the associated Comms Public key that this wallet is assigned. When encryption is applied to the wallet the Comms Secret Key is encrypted but the Comms Public Key is left in the clear. When a client starts up an encrypted wallet and doesn’t supply a passphrase it is easy to see that this Comms Secret Key cannot be decrypted. If the incorrect Passphrase is provided the Comms Secret Key also cannot be decrypted and cannot be compared to the Comms Public Key. If the correct Passphrase is provided then the Comms Secret Key can be decrypted, its associated Public Key can be derived and compared to the stored Comms Public key to confirm that they match. Then the Output Manager Service and Transaction Service databases are given the correct cipher so that they can perform decryption/encryption as required during normal operation.

@kukabi @Jasonvdb Please note the new parameter to the `wallet_create(...)` function for the optional passphrase. To retain current behaviour this parameter should be Null. There are also two new FFI function to apply and remove encryption for when you get to implementing this in the mobile clients.

## How Has This Been Tested?
Tests provided for the database backend abstractions and for the final integration into the wallet as a whole.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

sdbondi commented on base_layer/wallet/migrations/2020-06-29-130334_add_id_primary_key_to_outputs/down.sql:
```diff
@@ -0,0 +1,16 @@
+PRAGMA foreign_keys=off;
+BEGIN TRANSACTION;
```
Diesel already runs each migration in a transaction AFAIK (https://docs.rs/diesel_migrations/1.4.0/diesel_migrations/fn.run_pending_migrations.html). I seem to remember having to remove BEGIN TRANSACTION/COMMIT from my migrations or they would error.

philipr-za replied:
You are right, I took it out of the Up but forgot this one

sdbondi commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -114,23 +147,51 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
+            DbKey::UnspentOutputs => {
+                let mut outputs = OutputSql::index_status(OutputStatus::Unspent, &(*conn))?;
+
+                if let Some(cipher) = (*cipher).clone() {
+                    for o in outputs.iter_mut() {
+                        o.decrypt(&cipher)
+                            .map_err(|_| OutputManagerStorageError::AeadError("Decryption Error".to_string()))?;
+                    }
+                }
```
This should work (same for many other places) 
```suggestion
                if let Some(cipher) = cipher.as_ref() {
                    for o in outputs.iter_mut() {
                        o.decrypt(cipher)
                            .map_err(|_| OutputManagerStorageError::AeadError("Decryption Error".to_string()))?;
                    }
                }
```

philipr-za replied:
interesting. TIL

sdbondi commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -140,50 +201,99 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
+            DbKey::InvalidOutputs => {
+                let mut outputs = OutputSql::index_status(OutputStatus::Invalid, &(*conn))?;
+
+                if let Some(cipher) = (*cipher).clone() {
```
Feels like these calls should be DRY'd up a bit: Something like:

```rust
 fn decrypt_if_necessary<T: Encryptable<Aes256Gcm>>(&self, mut o: T) -> Result<T, OutputManagerStorageError> {
        if let Some(cipher) = self.cipher.as_ref() {
            o.decrypt(cipher)
                .map_err(|err| OutputManagerStorageError::AeadError("Decryption Error".to_string()))?;
        }
        Ok(o)
    }
```

philipr-za replied:
Nice, ja didn't think of that. Going to try do this in place rather than consuming the argument though but def much neater.

sdbondi commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -140,50 +201,99 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
                     }
-                    OutputSql::new(*o, OutputStatus::Spent, None).commit(&(*conn))?
+                    let mut new_output = NewOutputSql::new(*o, OutputStatus::Spent, None);
+                    if let Some(cipher) = (*cipher).clone() {
```
Encryption can also be DRY'd up

sdbondi commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -441,11 +572,75 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
     }
+
+    fn apply_encryption(&self, cipher: Aes256Gcm) -> Result<(), OutputManagerStorageError> {
+        let conn = acquire_lock!(self.database_connection);
```
Move this line below line 588 - no lock on the database is required if the function exits early

sdbondi commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -441,11 +572,75 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
+            .map_err(|_| OutputManagerStorageError::AeadError("Encryption Error".to_string()))?;
+        key_manager_state.set_state(&conn)?;
+
+        (*current_cipher) = Some(cipher);
```
It's a shame that this is the only reason that we need a RwLock (otherwise the cipher itself doesnt do any mutation so the lock wouldn't be needed at all). Sure there's a lock free way to set encryption at runtime but this is 👌  for now.

philipr-za replied:
Yup, couldn't really think of one that could be updated once the DB has started though.

sdbondi commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -27,27 +27,175 @@ use crate::{
+    {
+        let cipher = acquire_read_lock!(self.cipher);
+
+        match (*cipher).clone() {
```
```suggestion
        match cipher.as_ref() {
```

sdbondi commented on base_layer/wallet/src/util/encryption.rs:
```diff
@@ -0,0 +1,70 @@
+}
+
+pub fn encrypt_bytes_integral_nonce(cipher: &Aes256Gcm, plaintext: Vec<u8>) -> Result<Vec<u8>, AeadError> {
+    let mut nonce = [0u8; 12];
```
```suggestion
    let mut nonce = [0u8; AES_NONCE_BYTES];
```

sdbondi commented on base_layer/wallet/src/util/encryption.rs:
```diff
@@ -0,0 +1,70 @@
+}
+
+pub fn decrypt_bytes_integral_nonce(cipher: &Aes256Gcm, mut ciphertext: Vec<u8>) -> Result<Vec<u8>, AeadError> {
+    let data = ciphertext.split_off(AES_NONCE_BYTES);
+    let nonce = GenericArray::from_slice(ciphertext.as_slice());
+    cipher.decrypt(nonce, data.as_ref())
```
nit: a little cleaner?
```suggestion
  let (nonce_bytes, cipher_text) = ciphertext.split_at(AES_NONCE_BYTES);
    let nonce = GenericArray::from_slice(nonce_bytes);
    cipher.decrypt(nonce, cipher_text)
```

sdbondi replied:
Bit concerned that the input is not checked anywhere that I can see. `split_at` (and `split_off`) will panic if the input is no long enough, so you'd need to check either in the calling code or here for the correct data length

philipr-za replied:
Def, I remember putting it in there but must of fallen out along the way.

sdbondi reviewed (commented):
Nice! Like the clean tests - a couple of minor comments.

sdbondi commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -27,27 +27,175 @@ use crate::{
+        // CommsPublicKey that is stored in the db
+        {
+            let conn = acquire_lock!(database_connection);
+            let secret_key = WalletSettingSql::get(format!("{}", DbKey::CommsSecretKey), &conn)?;
+            let db_public_key = WalletSettingSql::get(format!("{}", DbKey::CommsPublicKey), &conn)?;
```
nit: using `to_string` instead of `format!` is more idiomatic
```suggestion
            let secret_key = WalletSettingSql::get(DbKey::CommsSecretKey.to_string()), &conn)?;
            let db_public_key = WalletSettingSql::get(DbKey::CommsPublicKey.to_string()), &conn)?;
```

sdbondi replied:
If you wanted to get fancy:

```rust
   pub fn get<T: ToString>(key: T, conn: &SqliteConnection) -> Result<Option<String>, WalletStorageError> {
        wallet_settings::table
            .filter(wallet_settings::key.eq(key.to_string()))
           (...)
    }
```
and then

```rust
let secret_key = WalletSettingSql::get(DbKey::CommsSecretKey), &conn)?;
```

sdbondi replied:
though, you probably just want `WalletSettingsSql` to only take in a `DbKey`. This way an invalid key will produce a compile error. Same for the other functions that take a key in.

philipr-za replied:
I am not sure I like the DbKey leaking from the WalletBackend trait implementation. WalletSettingSql and all the *Sql structs are direct views into the DB really. The logic of how the db is being used and its input/output validated live in the trait implementation and the WalletSqliteDatabase.

sdbondi replied:
It's all in the same `storage` module and internal to WalletBackend, so not sure I'd call that a leak. In any case, I think the code is neater if you use either DbKey directly or the `T: ToString` (either I believe would amount to the same binary output)

philipr-za replied:
I disagree about it not being a leak because if we wanted to just nix the whole WalletBackend trait abstraction for whatever reason we would have to come into this specialized Db interface struct and make a change.

sdbondi commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -63,7 +211,24 @@ impl WalletBackend for WalletSqliteDatabase {
             DbKey::CommsSecretKey => {
                 if let Some(key_str) = WalletSettingSql::get(format!("{}", key), &conn)? {
-                    Some(DbValue::CommsSecretKey(CommsSecretKey::from_hex(key_str.as_str())?))
+                    let secret_key = match (*cipher).clone() {
```
```suggestion
                    let secret_key = match cipher.as_ref() {
```

philipr-za replied:
ah missed this one.

sdbondi commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -101,11 +265,56 @@ impl WalletBackend for WalletSqliteDatabase {
+
+    fn apply_encryption(&self, cipher: Aes256Gcm) -> Result<(), WalletStorageError> {
+        let mut current_cipher = acquire_write_lock!(self.cipher);
+        if (*current_cipher).is_some() {
```
The `.` operator already does a dereference

```suggestion
        if current_cipher.is_some() {
```

philipr-za replied:
orly? TIL

sdbondi reviewed (dismissed):
Well done, gets my +1 - some minor nits that you are free to ignore

stringhandler commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -441,11 +539,75 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
+        }
+
+        let mut key_manager_state = KeyManagerStateSql::get_state(&conn)?;
+        key_manager_state
```
I might be mistaken, but it seems like if there are no outputs, the key_manager_state could get encrypted twice.

philipr-za replied:
This is true, lemme add the check quick.

stringhandler commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -547,14 +696,79 @@ impl OutputSql {
+            let ou = DbUnblindedOutput::try_from(output.clone())?;
+            diesel::update(outputs::table.filter(outputs::id.eq(&output.id)))
+                .set(outputs::hash.eq(Some(ou.hash.clone())))
+                .execute(conn)?;
```
Is there a transaction around this operation in the calling method?

philipr-za replied:
There isn't I have an open issue about adding transactions to the wallet database that I have not gotten to: https://github.com/tari-project/tari/issues/908

stringhandler reviewed (commented):
There's a lot of code here, so difficult to tell just from a code review if there is something missing. I'm happy for it to go in, just a few comments. This isn't enabled by default, so shouldn't break anything existing.

philipr-za committed:
Add partial wallet backup for unencrypted wallet backup

This PR adds functionality to produce a partial wallet backup that excludes the Comms Secret Key from the backup. This is to be used when the balance in a wallet is below a certain threshold so is low risk to backup without passphrase encryption. 

This use case is to allow for automatic backing up in the early use of the wallet while there are little funds so we don’t want to bother the nascent user with passphrase when the risk is low. When the funds are above a certain threshold an encryption passphrase would be required.

This PR adds:
- DB migration to store transaction direction for CompletedTransactions rather than relying on examining the Source Public Key. This is so that after a backup where the Private Key is cleared event though a new private key is generated the wallet can still tell which transactions were inbound or outbound.
- Functionality to perform the backup, clear the sensitive data and write it to the specified file.
- FFI interface to this functionality.

Restoration will be done by copying this backup file to the device and starting the wallet providing this db file as the database.

Because it is a Sqlite file and has the migration history inside it will automatically be migrated to new DB schema. When it is detected that there is no Comms Secret Key in the database a new one will be generated.
